cocotb/scheduler.py

```diff
@@ -578,8 +578,8 @@ def queue_function(self, coroutine):
         # each entry always has a unique thread.
         t, = matching_threads
 
-        t.thread_suspend()
         self._pending_coros.append(coroutine)
+        t.thread_suspend()
```

Comment:
Doesn't this now have the opposite problem where if `coroutine` ends too quickly, `t` is resumed before it is suspended?

Reply:
The coroutine is run in the main thread, which is stuck [here](https://github.com/cocotb/cocotb/blob/master/cocotb/scheduler.py#L757-L760) until `thread_suspend` is called. The function wrapper [waits for the coroutine to have run](https://github.com/cocotb/cocotb/blob/master/cocotb/decorators.py#L350) before calling `thread_resume`, so resume can't happen before suspend.

Reply:
That argument seems reasonable. Can you add a comment like
```python
# scheduler thread is blocked here and resumes when thread_suspend is called, so make sure everything
# is queued first
```

Reply:
Also, a test for the case above would be handy